src/Discord.Net.Rest/Net/DefaultRestClient.cs

```diff
@@ -101,62 +101,68 @@ public void SetCancelToken(CancellationToken cancelToken)
+#pragma warning restore IDISP004
+
+            if (reason != null)
+                restRequest.Headers.Add("X-Audit-Log-Reason", Uri.EscapeDataString(reason));
+            if (requestHeaders != null)
+                foreach (var header in requestHeaders)
+                    restRequest.Headers.Add(header.Key, header.Value);
+            var content = new MultipartFormDataContent("Upload----" + DateTime.Now.ToString(CultureInfo.InvariantCulture));
+
+            static StreamContent GetStreamContent(Stream stream)
             {
-                if (reason != null)
-                    restRequest.Headers.Add("X-Audit-Log-Reason", Uri.EscapeDataString(reason));
-                if (requestHeaders != null)
-                    foreach (var header in requestHeaders)
-                        restRequest.Headers.Add(header.Key, header.Value);
-                var content = new MultipartFormDataContent("Upload----" + DateTime.Now.ToString(CultureInfo.InvariantCulture));
-                MemoryStream memoryStream = null;
-                if (multipartParams != null)
+                if (stream.CanSeek)
+                {
+                    // Reset back to the beginning; it may have been used elsewhere or in a previous request.
+                    stream.Position = 0;
+                }
+
+#pragma warning disable IDISP004
+                return new StreamContent(stream);
+#pragma warning restore IDISP004
+            }
+
+            foreach (var p in multipartParams ?? ImmutableDictionary<string, object>.Empty)
+            {
+                switch (p.Value)
                 {
-                    foreach (var p in multipartParams)
-                    {
-                        switch (p.Value)
-                        {
 #pragma warning disable IDISP004
-                            case string stringValue:
-                                { content.Add(new StringContent(stringValue, Encoding.UTF8, "text/plain"), p.Key); continue; }
-                            case byte[] byteArrayValue:
-                                { content.Add(new ByteArrayContent(byteArrayValue), p.Key); continue; }
-                            case Stream streamValue:
-                                { content.Add(new StreamContent(streamValue), p.Key); continue; }
-                            case MultipartFile fileValue:
-                                {
-                                    var stream = fileValue.Stream;
-                                    if (!stream.CanSeek)
-                                    {
-                                        memoryStream = new MemoryStream();
-                                        await stream.CopyToAsync(memoryStream).ConfigureAwait(false);
-                                        memoryStream.Position = 0;
-#pragma warning disable IDISP001
-                                        stream = memoryStream;
-#pragma warning restore IDISP001
-                                    }
```

Comment:
I got rid of this whole `MemoryStream` logic intentionally because I don't think it gains us anything.

If the request is only performed once, then the initial `Stream` will be read fully and we wouldn't have benefited from first copying it into a `MemoryStream`.

If the integrating app tries second request with the same stream, or the first request fails with a 429 and is retried, then copying the stream a second time won't be possible. In other words, if the incoming stream can't be seeked, then it can't be reused for future requests, and we're gonna be in a pickle here no matter what.